mediatek: add support for Ruijie RG-EW3200GX PRO

1248289414 opened the pull request:
X32 Pro is another product name for it in the Chinese market.

Specifications:
- SoC: MT7622B
- RAM: 256MB
- Flash: XMC XM25QH128C or Winbond WQ25Q128JVSQ 16MB SPI NOR
- Ethernet: 5x1GbE
- Switch: MT7531BE
- WiFi: 2.4G: MT7622 5G: MT7915AN+MT7975AN
- 3LEDs: System LED (blue) + Mesh LED (green) + Mesh LED (red)
- 2Keys: Mesh button + Reset button
- UART: Marked J19 on board. 3.3v, 115200n1
- Power: 12V 2.5A

MAC addresses as verified by OEM firmware:
use   address    source
WAN   *: F4       ethaddr@product_info
LAN    *: F5
5g       *: F6
2g       *: F7

Flash instruction:
1. Serve the initramfs.img using a TFTP server with address 10.10.10.3.
2. Interrupt the uboot startup process via UART.
3. Select "System Load Linux to SDRAM via TFTP" item.
4. (important) Back up firmware(mtd7) partitions with:
        dd if=/dev/mtd7 of=/tmp/firmware.bin
   and then download the firmware.bin image via SCP.
5. Flash the OpenWrt sysupgrade firmware.

Recovery stock firmware:
1. Transfer the firmware.bin image to the device.
2. Flash the image with:
        mtd write firmware.bin firmware

------------

The XMC XM25QH128C is a 16MB SPI NOR chip. Datasheet available at https://www.xmcwh.com/uploads/435/XM25QH128C.pdf

Signed-off-by: Langhua Ye <y1248289414@outlook.com>

hauke commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,317 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&epa_elna_pins>;
+	mediatek,mtd-eeprom = <&factory 0x0>;
+};
```
Please add an empty line to the end.

1248289414 replied:
Fixed✔

hauke commented on target/linux/mediatek/mt7622/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -0,0 +1,18 @@
+	ruijie,rg-ew3200gx-pro)
+		hw_mac_addr=$(mtd_get_mac_ascii product_info ethaddr)
+		[ "$PHYNBR" = "0" ] && macaddr_add $hw_mac_addr 3 > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && macaddr_add $hw_mac_addr 2 > /sys${DEVPATH}/macaddress
```
@blocktrron  and @adschm  Do we have no better way to assign MACs from the U-Boot env to wifi devices?

dangowrt replied:
Yes, by setting `ethernet0 = &gmac0;` as alias in the DTS, then U-Boot will patch fdt to pass it down to Linux.

hauke commented on target/linux/mediatek/patches-5.10/421-mtd-spi-nor-add-xmc-xm25qh128c.patch:
```diff
@@ -0,0 +1,11 @@
+ 	{ "XM25QH128A", INFO(0x207018, 0, 64 * 1024, 256,
+ 			     SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
++	{ "XM25QH128C", INFO(0x204018, 0, 64 * 1024, 256,
++			     SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
```
This change looks good. Please send this patch also to the upstream Linux kernel mailing list at linux-mtd@lists.infradead.org to get it integrated into mainline Linux kernel. Move this to the target/linux/generic/pending-5.10/ folder in OpenWrt, it is likely that more device will use this flash chip in the future.

1248289414 replied:
Done, I will send it to the upstream later.

1248289414 commented:
UPDATE:  Add memory nodes to comply with DTS rules.

dangowrt commented:
Please split-off the part which adds the XM25QH128C flash chip to `target/linux/generic` into a separate commit with it's own description, applied before the main commit adding support for the Ruijie device.
In this way we can send the patch adding the flash chip upstream (unless nobody else did already).

1248289414 commented:
I have split this PR into two commits. @dangowrt

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,321 @@
+		led-failsafe = &led_system;
+		led-running = &led_system;
+		led-upgrade = &led_system;
+		serial0 = &uart0;
```
Please add `ethernet0` alias as well because then U-boot will pass down the MAC address to Linux and there is no need to set it up in userspace (`wan` MAC address will still need handling in userspace, of course)

1248289414 replied:
@dangowrt Thanks for the suggestion. I tested it and it works well.

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:
```diff
@@ -29,9 +30,30 @@ mediatek_setup_interfaces()
+
+	case $board in
+	ruijie,rg-ew3200gx-pro)
+		wan_mac=$(mtd_get_mac_ascii product_info ethaddr)
```
See comment in DTS how to make this a bit better (by setting `ethernet0` alias in DTS instead).

1248289414 commented:
UPDATE: Add `ethernet0 = &gmac0; `

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,323 @@
+	};
+
+	memory {
+		reg = <0 0x40000000 0 0xf000000>;
```
Using `0xf000000` (== 240M) looks weird to me and memory autodetection is working fine for other devices with a total of 256M. Is that necessary due to use of the upper 16M by something else (ie. TF-A or OP-TEE)? If so, a `reserved-memory` node would be a more appropriate way to express that.

1248289414 replied:
I copied the `memory` nodes from the original DTS directly. Memory auto-recognition worked fine for this device. Maybe I can delete the `memory` node?

dangowrt replied:
I see that other boards have
```
reg = <0 0x40000000 0 0x40000000>;
```
which should enable memory size auto-detection with the full addressable range of the CPU. As this is the same for all boards, we could as well remove it from the board-specific DTS and move it to `mt7622.dtsi` instead.
Anyway, you will have to try and see if it works (ie. boots successfully with full amount of memory detected and usable, no overlap with ranges used by TF-A or OP-TEE)

1248289414 replied:
I changed the range in `memory` node. It was tested and works fine. The range of memory automatically detected has the same size as the original image. 
It looks like that the range of memory used by TF-A has been included in [mt7622.dtsi](https://github.com/torvalds/linux/blob/fb184c4af9b9f4563e7a126219389986a71d5b5b/arch/arm64/boot/dts/mediatek/mt7622.dtsi#L130)😀.

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,323 @@
+			partition@B0000 {
+				label = "u-boot-env";
+				reg = <0xb0000 0x20000>;
+				read-only;
```
I reckon that `u-boot-env` can be accessed using `uboot-envtools`. Maybe also add the board there (in an additional commit) and remove the `read-only` attribute.

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:
```diff
@@ -29,9 +30,28 @@ mediatek_setup_interfaces()
+
+	case $board in
+	ruijie,rg-ew3200gx-pro)
+		lan_mac=$(macaddr_add $(mtd_get_mac_ascii u-boot-env ethaddr) 1)
```
Now that WAN and label mac address are passed down by U-Boot (because the `ethernet0` alias is now present), I think it would be nicer to just read the address already assigned to `eth0` at this point instead of reading it from flash.

1248289414 replied:
Would using `$(macaddr_add $(get_mac_label) 1) ` make it more readable?

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:
```diff
@@ -29,9 +30,28 @@ mediatek_setup_interfaces()
+	esac
+
+	[ -n "$lan_mac" ] && ucidef_set_interface_macaddr "lan" $lan_mac
+	[ -n "$wan_mac" ] && ucidef_set_interface_macaddr "wan" $wan_mac
```
This line is now unused and can be removed. Once there is a board which needs it, we can easily add it again.

1248289414 replied:
✅Removed.

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:
```diff
@@ -29,9 +30,28 @@ mediatek_setup_interfaces()
+
+	[ -n "$lan_mac" ] && ucidef_set_interface_macaddr "lan" $lan_mac
+	[ -n "$wan_mac" ] && ucidef_set_interface_macaddr "wan" $wan_mac
+	[ -n "$label_mac" ] && ucidef_set_label_macaddr $label_mac
```
Same here.

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -0,0 +1,18 @@
+
+case "$board" in
+	ruijie,rg-ew3200gx-pro)
+		hw_mac_addr=$(mtd_get_mac_ascii u-boot-env ethaddr)
```
No need to read the address from flash now that it gets passed down to Linux by U-Boot.

1248289414 replied:
`[ "$PHYNBR" = "0" ] && macaddr_add $(get_mac_label) 3 > /sys${DEVPATH}/macaddress`
`[ "$PHYNBR" = "1" ] && macaddr_add $(get_mac_label) 2 > /sys${DEVPATH}/macaddress`
Is it right?

dangowrt replied:
Yes, that looks good.

dangowrt commented:
Please squash the merge-commit (which was needed due to conflict in `02_network`) and then we are good to go I think.

1248289414 committed:
kernel: add support for XMC XM25QH128C

The XMC XM25QH128C is a 16MB SPI NOR chip. The patch is verified on Ruijie RG-EW3200GX PRO.
Datasheet available at https://www.xmcwh.com/uploads/435/XM25QH128C.pdf

Signed-off-by: Langhua Ye <y1248289414@outlook.com>

1248289414 commented:
✅Done. @dangowrt

dangowrt commented:
> WiFi: 2.4G: MT7622 5G: MT7975A+MT7615N
The 5G combination of chips isn't plausible. You probably meant MT7975A+MT7**9**15N... We also typically denote the chips in order of being connected to the SoC, ie. MT7915N is connected via PCIe to the SoC and then the RF frontend MT7975A is connected to that, so better write `MT7915N+MT7975A`.

1248289414 commented:
>  WiFi: 2.4G: MT7622 5G: MT7975A+MT7615N

Oh, this is a mistake.😥 This device supports AX and the 5G is provided by `MT7915AN+MT7975AN`. I have changed the commit message. @dangowrt

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,322 @@
+		switch@0 {
+			compatible = "mediatek,mt7531";
+			reg = <0>;
+			reset-gpios = <&pio 54 GPIO_ACTIVE_HIGH>;
```
Sorry to be pita and come find new things with every iteration. One last thing:
Can you try adding
```
interrupt-controller;
#interrupt-cells = <1>;
interrupt-parent = <&pio>;
interrupts = <53 IRQ_TYPE_LEVEL_HIGH>;
```
and check if link-status reporting still works as expected?
I assume that GPIO#53 is connected as interrupt line to the MT7531 switch just like on all other MT7622+MT7531 boards, and declaring it in device tree allow to detect link status changes using interrupts instead of having to poll.

1248289414 replied:
Your assumption is correct. The link-status can be properly reported after adding and incorrect value will produce error logs. But the link-status can also be properly reported without this part. Should I add it? @dangowrt

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,327 @@
+};
+
+&slot0 {
+	mt7615@0,0 {
```
Just cosmetics: The node should be called `mt7915` or just `mt76`. Calling it `mt7615` is a bit misleading...

1248289414 replied:
✔.

dangowrt commented on target/linux/mediatek/dts/mt7622-ruijie-rg-ew3200gx-pro.dts:
```diff
@@ -0,0 +1,327 @@
+			};
+
+			partition@B0000 {
+				label = "u-boot-env";
```
Please also add an additional commit adding the board to uboot-envtools (`package/boot/uboot-envtools/files/mediatek_mt7622`)

1248289414 replied:
Finish! @dangowrt

1248289414 committed:
mediatek: add support for Ruijie RG-EW3200GX PRO

X32 Pro is another product  name for it in the Chinese market.

Specifications:
- SoC: MT7622B
- RAM: 256MB
- Flash: XMC XM25QH128C or Winbond WQ25Q128JVSQ 16MB SPI NOR
- Ethernet: 5x1GbE
- Switch: MT7531BE
- WiFi: 2.4G: MT7622 5G: MT7915AN+MT7975AN
- 3LEDs: System LED(blue) + Mesh LED(green) + Mesh LED(red)
- 2Keys: Mesh button + Reset button
- UART: Marked J19 on board. 3.3v, 115200n1
- Power: 12V 2.5A

MAC addresses as verified by OEM firmware:
use   address    source
WAN   *:F4       ethaddr@product_info
LAN   *:F5
5g    *:F6
2g    *:F7

Flash instruction:
1. Serve the initramfs.img using a TFTP server with address 10.10.10.3.
2. Interrupt the uboot startup process via UART.
3. Select "System Load Linux to SDRAM via TFTP" item.
4. (important) Back up firmware(mtd7) partitions with:
        dd if=/dev/mtd7 of=/tmp/firmware.bin
   and then download the firmware.bin image via SCP.
5. Flash the OpenWrt sysupgrade firmware.

Recovery stock firmware:
1. Transfer the firmware.bin image to the device.
2. Flash the image with:
        mtd write firmware.bin firmware

Signed-off-by: Langhua Ye <y1248289414@outlook.com>

dangowrt reviewed (changes requested):
Please also add patch description for commit `uboot-envtools: mt7622: add support for Ruijie RG-EW3200GX PRO`. Just the commit title is not enough.
Suggestion:
```
uboot-envtools: mt7622: add support for Ruijie RG-EW3200GX PRO

Add U-Boot environment settings for Ruijie RG-EW3200GX PRO to allow
users to access the bootloader environment using fw_printenv/fw_setenv
while running OpenWrt.

Signed-off-by: Langhua Ye <y1248289414@outlook.com>
```

1248289414 committed:
uboot-envtools: mt7622: add support for Ruijie RG-EW3200GX PRO

Add U-Boot environment settings for Ruijie RG-EW3200GX PRO to allow
users to access the bootloader environment using fw_printenv/fw_setenv
while running OpenWrt.

Signed-off-by: Langhua Ye <y1248289414@outlook.com>

dangowrt commented:
@hauke can you commit this to git.openwrt.org? I'm on vacation and don't have my development machine with me...

1248289414 commented:
Great. Thank you for your vacation time. 😀

hauke commented:
Thank you for your patches, I applied them to master in df1383f7964527b2e7178ae744cd167a2c230f9f, ce8a33b021a703b1933f627835f6cdc88c0ecdd3 and d15f9b9043eac41ecfa1aaf1d65dca460b3c7fb4.
